File: generate-parentheses/generate-parentheses.cpp
class Solution {
    void solver(vector<string> &result, string str, int OpenCount, int CloseCount, int maxSize){
        
        if(str.size() == 2*maxSize){
            result.push_back(str);
            return;
        }
        
        if(OpenCount < maxSize) solver(result, str + "(", OpenCount + 1, CloseCount, maxSize);
        if(CloseCount < OpenCount) solver(result, str + ")", OpenCount, CloseCount + 1, maxSize);
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        solver(result, "", 0, 0, n);
        return result;
    }
};

File: generate-parentheses/generate-parentheses.java
class Solution {
    public void helper(int n,int openCount,int closeCount, String str, List<String> list ){
        if(openCount == n && closeCount == n){
            list.add(str);
            return ;
        }
        if(openCount < n){
            helper(n, openCount + 1, closeCount, str + "(", list);
        }
        if(closeCount < openCount){
            helper(n, openCount, closeCount + 1, str + ")", list);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> list = new ArrayList<>();
        helper(n, 0, 0, "", list);
        return list;
    }
}